pkg/proxy/proxy.go

```diff
@@ -315,22 +323,33 @@ func ProxyHandler(config *cfg.Config, logger log.Logger) func(http.ResponseWrite
 
 // Forward the first valid response for non-query endpoints
 func forwardFirstResponse(w http.ResponseWriter, results <-chan *http.Response, logger log.Logger) {
+	forwarded := false
 	for resp := range results {
-		// Directly copy all headers and body from Loki response to Grafana
-		for key, values := range resp.Header {
-			for _, value := range values {
-				w.Header().Add(key, value)
+		if !forwarded {
+			// Forward the first response
+			for key, values := range resp.Header {
+				for _, value := range values {
+					w.Header().Add(key, value)
+				}
+			}
+
+			w.Header().Set("Connection", "keep-alive")
+			w.WriteHeader(resp.StatusCode)
+			if _, err := io.Copy(w, resp.Body); err != nil {
+				level.Error(logger).Log("msg", "Failed to copy response body", "err", err)
+			}
+			forwarded = true
+		} else {
+			// Drain the body of non-forwarded responses to prevent connection leaks
+			if _, err := io.Copy(io.Discard, resp.Body); err != nil {
+				level.Error(logger).Log("msg", "Failed to drain response body", "err", err)
 			}
 		}
 
-		w.Header().Set("Connection", "keep-alive")
-		w.WriteHeader(resp.StatusCode)
-		_, err := io.Copy(w, resp.Body) // Forward the body as-is
-		if err != nil {
-			level.Error(logger).Log("msg", "Failed to copy response body", "err", err)
-			return
+		// Close all response bodies to prevent resource leaks
+		if err := resp.Body.Close(); err != nil {
+			level.Error(logger).Log("msg", "Failed to close response body", "err", err)
```

Comment:
Hi, just wandering if we should add a check in case there are no healthy upstreams and `forwarded` stays false, should we have a http error, wdyt?